src/crimson/osd/pg.cc

```diff
@@ -43,933 +41,211 @@ namespace {
-       boost::adaptors::filtered(std::not_fn(is_enough)) |
-       boost::adaptors::filtered(std::not_fn(has_covered)) |
-       boost::adaptors::filtered(is_good)),
-      cands_sorted_by_eversion_t{},
-      [&](cands_sorted_by_eversion_t& cands, int osd) {
-        cands.emplace(get_info(osd).last_update, get_shard(osd));
-        return cands;
-      });
-    boost::copy(cands |
-                boost::adaptors::map_values |
-                boost::adaptors::transformed(shard_to_osd),
-                std::back_inserter(want));
-  }
-  return want;
-}
-
-bool PG::proc_replica_info(pg_shard_t from,
-                           const pg_info_t& pg_info,
-                           epoch_t send_epoch)
-{
-
-  if (auto found = peer_info.find(from);
-      found != peer_info.end() &&
-      found->second.last_update == pg_info.last_update) {
-    logger().info("got info {} from osd.{}, identical to ours",
-                  info, from);
-    return false;
-  } else if (!osdmap->has_been_up_since(from.osd, send_epoch)) {
-    logger().info("got info {} from down osd.{}. discarding",
-                  info, from);
-    return false;
-  } else {
-    logger().info("got info {} from osd.{}", info, from);
-    peer_info.emplace(from, pg_info);
-    return true;
-  }
-}
-
-void PG::proc_replica_log(pg_shard_t from,
-                          const pg_info_t& pg_info,
-                          const pg_log_t& pg_log,
-                          const pg_missing_t& pg_missing)
-{
-
-  logger().info("{} for osd.{}: {} {} {}", from, pg_info, pg_log, pg_missing);
-  peer_info.insert_or_assign(from, pg_info);
-}
-
-// Returns an iterator to the best info in infos sorted by:
-//  1) Prefer newer last_update
-//  2) Prefer longer tail if it brings another info into contiguity
-//  3) Prefer current primary
-pg_shard_t
-PG::find_best_info(const PG::peer_info_t& infos) const
-{
-  // See doc/dev/osd_internals/last_epoch_started.rst before attempting
-  // to make changes to this process.  Also, make sure to update it
-  // when you find bugs!
-  auto min_last_update_acceptable = eversion_t::max();
-  epoch_t max_les = 0;
-  for ([[maybe_unused]] auto& [shard, info] : infos) {
-    if (max_les < info.history.last_epoch_started) {
-      max_les = info.history.last_epoch_started;
-    }
-    if (!info.is_incomplete() &&
-        max_les < info.last_epoch_started) {
-      max_les = info.last_epoch_started;
-    }
-  }
-  for ([[maybe_unused]] auto& [shard, info] : infos) {
-    if (max_les <= info.last_epoch_started &&
-        min_last_update_acceptable > info.last_update) {
-      min_last_update_acceptable = info.last_update;
-    }
-  }
-  if (min_last_update_acceptable == eversion_t::max()) {
-    return pg_shard_t{};
-  }
-  // find osd with newest last_update (oldest for ec_pool).
-  // if there are multiples, prefer
-  //  - a longer tail, if it brings another peer into log contiguity
-  //  - the current primary
-  struct is_good {
```

Comment:
instead of completely dropping the code, i'd like at least port part of it to the `PeeringState.cc` for better readability once this change is merged.

Reply:
Sure, this is probably easier to read than what's in there.